Java 14 records with empty constructors violate beRecords() rule

Given that I have an arch unit test like this:

```
package org.example;

import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;

import com.tngtech.archunit.core.importer.ClassFileImporter;
import org.junit.jupiter.api.Test;

public class RecordTest
{

  @Test
  public void recordTest() {
    var javaClasses = new ClassFileImporter().importPackages("org.example");

    classes().that()
        .haveNameMatching("(.*)Record")
        .should()
        .beRecords()
        .because("they should be simple data structures")
        .check(javaClasses);
  }
}

```

And a record like this in my project:

```
package org.example;

public record EmptyRecord()
{
}
```

When I run the arch unit test, then a violation is reported, even though `EmptyRecord` is a valid (but useless) Java 14 record:

```
Architecture Violation [Priority: MEDIUM] - Rule 'classes that have name matching '(.*)Record' should be records, because they should be simple data structures' was violated (1 times):
Class <org.example.EmptyRecord> is no record in (EmptyRecord.java:0)
java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'classes that have name matching '(.*)Record' should be records, because they should be simple data structures' was violated (1 times):
Class <org.example.EmptyRecord> is no record in (EmptyRecord.java:0)
	at com.tngtech.archunit.lang.ArchRule$Assertions.assertNoViolation(ArchRule.java:94)
	at com.tngtech.archunit.lang.ArchRule$Assertions.check(ArchRule.java:86)
	at com.tngtech.archunit.lang.ArchRule$Factory$SimpleArchRule.check(ArchRule.java:165)
	at org.example.RecordTest.recordTest(RecordTest.java:20)
```

When the record is extended with a parameter, then the violation is resolved:

```
package org.example;

public record EmptyRecord(String someValue)
{
}
```

A full example to reproduce the issue can be found [here](https://github.com/SoylentBob/arch-unit-record-bug).

--- a/archunit-integration-test/src/test/java/com/tngtech/archunit/ImporterRules.java
+++ b/archunit-integration-test/src/test/java/com/tngtech/archunit/ImporterRules.java
@@ -9,8 +9,10 @@ import com.tngtech.archunit.junit.ArchTest;
 import com.tngtech.archunit.lang.ArchRule;
 
 import static com.tngtech.archunit.ArchUnitArchitectureTest.THIRDPARTY_PACKAGE_IDENTIFIER;
+import static com.tngtech.archunit.base.DescribedPredicate.doNot;
 import static com.tngtech.archunit.base.DescribedPredicate.not;
 import static com.tngtech.archunit.core.domain.JavaClass.Predicates.belongToAnyOf;
+import static com.tngtech.archunit.core.domain.JavaClass.Predicates.resideOutsideOfPackage;
 import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.noClasses;
 
 public class ImporterRules {
@@ -18,7 +20,22 @@ public class ImporterRules {
     @ArchTest
     public static final ArchRule domain_does_not_access_importer =
             noClasses().that().resideInAPackage("..core.domain..")
-                    .should().accessClassesThat(belong_to_the_import_context());
+                    .should().dependOnClassesThat(belong_to_the_import_context());
+
+    @ArchTest
+    public static final ArchRule asm_is_only_used_in_importer_or_JavaClassDescriptor =
+            noClasses()
+                    .that(
+                            resideOutsideOfPackage("..core.importer..")
+                                    .and(doNot(belongToAnyOf(JavaClassDescriptor.class
+                                            // Conceptually there are also dependencies from JavaModifier to ASM.
+                                            // However, at the moment all those are inlined by the compiler (primitives).
+                                            // Whenever we get the chance to break the public API
+                                            // we should remove the dependencies from JavaModifier to ASM.
+                                            // Those dependencies crept in by accident at some point because the design was convenient.
+                                    )))
+                    )
+                    .should().dependOnClassesThat().resideInAPackage("org.objectweb..");
 
     @ArchTest
     public static final ArchRule ASM_type_is_only_accessed_within_JavaClassDescriptor_or_JavaClassDescriptorImporter =

--- a/archunit/src/jdk16test/java/com/tngtech/archunit/core/importer/ClassFileImporterRecordsTest.java
+++ b/archunit/src/jdk16test/java/com/tngtech/archunit/core/importer/ClassFileImporterRecordsTest.java
@@ -4,21 +4,33 @@ import com.tngtech.archunit.core.domain.JavaClass;
 import com.tngtech.archunit.core.domain.JavaConstructor;
 import com.tngtech.archunit.core.domain.JavaField;
 import com.tngtech.archunit.core.domain.JavaMethod;
+import com.tngtech.java.junit.dataprovider.DataProvider;
+import com.tngtech.java.junit.dataprovider.DataProviderRunner;
+import com.tngtech.java.junit.dataprovider.UseDataProvider;
 import org.junit.Test;
+import org.junit.runner.RunWith;
 
 import static com.tngtech.archunit.testutil.Assertions.assertThat;
+import static com.tngtech.java.junit.dataprovider.DataProviders.testForEach;
 
+@RunWith(DataProviderRunner.class)
 public class ClassFileImporterRecordsTest {
+    @DataProvider
+    public static Object[][] imports_record() {
+        record SimpleRecord(String component1, int component2) {
+        }
+        record EmptyRecord() {
+        }
+        return testForEach(SimpleRecord.class, EmptyRecord.class);
+    }
 
     @Test
-    public void imports_simple_record() {
-        record RecordToImport(String component1, int component2) {
-        }
-
-        JavaClass javaClass = new ClassFileImporter().importClasses(RecordToImport.class, Record.class).get(RecordToImport.class);
+    @UseDataProvider
+    public void imports_record(Class<?> recordToImport) {
+        JavaClass javaClass = new ClassFileImporter().importClasses(recordToImport, Record.class).get(recordToImport);
 
         assertThat(javaClass)
-                .matches(RecordToImport.class)
+                .matches(recordToImport)
                 .hasRawSuperclassMatching(Record.class)
                 .hasNoInterfaces()
                 .isInterface(false)

--- a/archunit/src/main/java/com/tngtech/archunit/core/domain/JavaModifier.java
+++ b/archunit/src/main/java/com/tngtech/archunit/core/domain/JavaModifier.java
@@ -18,6 +18,7 @@ package com.tngtech.archunit.core.domain;
 import java.util.EnumSet;
 import java.util.Set;
 
+import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Sets;
 import com.tngtech.archunit.PublicAPI;
 import org.objectweb.asm.Opcodes;
@@ -73,7 +74,17 @@ public enum JavaModifier {
     @Deprecated
     @PublicAPI(usage = ACCESS)
     public static Set<JavaModifier> getModifiersForClass(int asmAccess) {
-        return getModifiersFor(ApplicableType.CLASS, asmAccess);
+        Set<JavaModifier> modifiers = getModifiersFor(ApplicableType.CLASS, asmAccess);
+        boolean opCodeForRecordIsPresent = (asmAccess & Opcodes.ACC_RECORD) != 0;
+        if (opCodeForRecordIsPresent) {
+            // As records are implicitly static and final (compare JLS 8.10 Record Declarations),
+            // we ensure that those modifiers are always present. (asmAccess does not contain STATIC.)
+            return ImmutableSet.<JavaModifier>builder()
+                    .addAll(modifiers)
+                    .add(JavaModifier.STATIC, JavaModifier.FINAL)
+                    .build();
+        }
+        return modifiers;
     }
 
     /**

--- a/archunit/src/main/java/com/tngtech/archunit/core/domain/JavaModifier.java
+++ b/archunit/src/main/java/com/tngtech/archunit/core/domain/JavaModifier.java
@@ -64,16 +64,37 @@ public enum JavaModifier {
         this.asmAccessFlag = asmAccessFlag;
     }
 
+    /**
+     * @deprecated This seems like an unnecessary API for users of ArchUnit, but limits us to do internal refactorings.
+     * If you think you need this API, please reach out to us on GitHub by creating an issue at
+     * <a href="https://github.com/TNG/ArchUnit/issues">https://github.com/TNG/ArchUnit/issues</a>.
+     * Otherwise, at some point in the future we will remove this API without any replacement.
+     */
+    @Deprecated
     @PublicAPI(usage = ACCESS)
     public static Set<JavaModifier> getModifiersForClass(int asmAccess) {
         return getModifiersFor(ApplicableType.CLASS, asmAccess);
     }
 
+    /**
+     * @deprecated This seems like an unnecessary API for users of ArchUnit, but limits us to do internal refactorings.
+     * If you think you need this API, please reach out to us on GitHub by creating an issue at
+     * <a href="https://github.com/TNG/ArchUnit/issues">https://github.com/TNG/ArchUnit/issues</a>.
+     * Otherwise, at some point in the future we will remove this API without any replacement.
+     */
+    @Deprecated
     @PublicAPI(usage = ACCESS)
     public static Set<JavaModifier> getModifiersForField(int asmAccess) {
         return getModifiersFor(ApplicableType.FIELD, asmAccess);
     }
 
+    /**
+     * @deprecated This seems like an unnecessary API for users of ArchUnit, but limits us to do internal refactorings.
+     * If you think you need this API, please reach out to us on GitHub by creating an issue at
+     * <a href="https://github.com/TNG/ArchUnit/issues">https://github.com/TNG/ArchUnit/issues</a>.
+     * Otherwise, at some point in the future we will remove this API without any replacement.
+     */
+    @Deprecated
     @PublicAPI(usage = ACCESS)
     public static Set<JavaModifier> getModifiersForMethod(int asmAccess) {
         return getModifiersFor(ApplicableType.METHOD, asmAccess);

--- a/archunit/src/main/java/com/tngtech/archunit/core/importer/JavaClassProcessor.java
+++ b/archunit/src/main/java/com/tngtech/archunit/core/importer/JavaClassProcessor.java
@@ -27,7 +27,6 @@ import java.util.stream.Stream;
 
 import com.google.common.collect.HashMultimap;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.SetMultimap;
 import com.google.common.primitives.Booleans;
 import com.google.common.primitives.Bytes;
@@ -58,7 +57,6 @@ import org.objectweb.asm.Handle;
 import org.objectweb.asm.Label;
 import org.objectweb.asm.MethodVisitor;
 import org.objectweb.asm.Opcodes;
-import org.objectweb.asm.RecordComponentVisitor;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -114,6 +112,7 @@ class JavaClassProcessor extends ClassVisitor {
         boolean opCodeForInterfaceIsPresent = (access & Opcodes.ACC_INTERFACE) != 0;
         boolean opCodeForEnumIsPresent = (access & Opcodes.ACC_ENUM) != 0;
         boolean opCodeForAnnotationIsPresent = (access & Opcodes.ACC_ANNOTATION) != 0;
+        boolean opCodeForRecordIsPresent = (access & Opcodes.ACC_RECORD) != 0;
         Optional<String> superclassName = getSuperclassName(superName, opCodeForInterfaceIsPresent);
         LOG.trace("Found superclass {} on class '{}'", superclassName.orElse(null), name);
 
@@ -123,7 +122,8 @@ class JavaClassProcessor extends ClassVisitor {
                 .withInterface(opCodeForInterfaceIsPresent)
                 .withEnum(opCodeForEnumIsPresent)
                 .withAnnotation(opCodeForAnnotationIsPresent)
-                .withModifiers(JavaModifier.getModifiersForClass(access));
+                .withModifiers(JavaModifier.getModifiersForClass(access))
+                .withRecord(opCodeForRecordIsPresent);
 
         className = descriptor.getFullyQualifiedClassName();
         declarationHandler.onNewClass(className, superclassName, interfaceNames);
@@ -151,21 +151,6 @@ class JavaClassProcessor extends ClassVisitor {
         if (!importAborted() && source != null) {
             javaClassBuilder.withSourceFileName(source);
         }
-    }
-
-    @Override
-    public RecordComponentVisitor visitRecordComponent(String name, String descriptor, String signature) {
-        javaClassBuilder.withRecord(true);
-
-        // Records are implicitly static and final (compare JLS 8.10 Record Declarations)
-        // Thus we ensure that those modifiers are always present (the access flag in visit(..) does not contain STATIC)
-        ImmutableSet<JavaModifier> recordModifiers = ImmutableSet.<JavaModifier>builder()
-                .addAll(javaClassBuilder.getModifiers())
-                .add(JavaModifier.STATIC, JavaModifier.FINAL)
-                .build();
-        javaClassBuilder.withModifiers(recordModifiers);
-
-        return super.visitRecordComponent(name, descriptor, signature);
     }
 
     @Override

